File: extension/connection/recovery-ladder.js
/**
 * @file recovery-ladder.js
 * @description Sequential, cancelable promise-chain that runs a list of
 * recovery "rungs" until one succeeds or all fail.
 *
 * Used by {@link ConnectionAuthority} to recover from `peerHealth = FAILED`
 * or `selfState = RECONNECTING`. The ladder owns sequencing, logging, and
 * cancellation; each rung's `action` owns its own success-signal subscription
 * and budget-timer (via the helpers in this module).
 *
 * Contract per rung:
 *   - `name`     : short identifier used in logs (`rung1`, `rung2`, ...).
 *   - `budgetMs` : informational — passed to the action so it knows its
 *                  budget. Logging includes per-rung elapsed time.
 *   - `action`   : `async ({ signal }) => boolean` — resolves `true` if the
 *                  rung recovered, `false` if it timed out / surrendered.
 *                  The provided `signal` is an AbortSignal-like object whose
 *                  `aborted` flag flips and `addEventListener('abort', ...)`
 *                  fires when {@link RecoveryLadder#cancel} is called. Actions
 *                  MUST observe this signal to tear down their subscriptions
 *                  and timers cleanly — otherwise cancellation leaks.
 *
 * Ladder semantics:
 *   - Rungs run strictly sequentially. The first `action` returning `true`
 *     ends the ladder with `{ ok: true, rung: name }`.
 *   - A rung returning `false` (or throwing) advances to the next rung.
 *   - If `cancel()` is called, the in-flight action is signalled and the
 *     ladder resolves with `{ ok: false, cancelled: true }` after the action
 *     settles. No further rungs are attempted.
 *   - If every rung fails, the last rung's outcome is reported as
 *     `{ ok: false, exhausted: true, lastError? }`.
 *
 * Logging:
 *   - `console.log('[CA] ladder: rung=N start')` on entry.
 *   - `console.log('[CA] ladder: rung=N ok t=X.Xs')`  on success.
 *   - `console.log('[CA] ladder: rung=N timeout t=X.Xs')` on action returning false.
 *   - `console.log('[CA] ladder: rung=N error t=X.Xs <message>')` on throw.
 *   - `console.log('[CA] ladder: cancelled')` once cancel-driven exit completes.
 *
 * Spec: docs/superpowers/specs/2026-05-02-connection-authority-design.md
 *       §"Recovery ladder".
 */

/**
 * @typedef {object} RecoveryRung
 * @property {string} name
 * @property {number} budgetMs
 * @property {(args: { signal: AbortSignal }) => Promise<boolean>} action
 */

/**
 * @typedef {object} LadderResult
 * @property {boolean} ok
 * @property {string} [rung]        - Name of the rung that succeeded (when ok=true).
 * @property {boolean} [cancelled]  - True when cancel() drove the exit.
 * @property {boolean} [exhausted]  - True when every rung returned false / threw.
 * @property {Error}   [lastError]  - Last action error, if any (when ok=false).
 */

export class RecoveryLadder {
  /**
   * @param {object} opts
   * @param {RecoveryRung[]} opts.rungs
   *   Ordered list of rungs to attempt. Must be non-empty.
   * @param {(transition: { rung: string, phase: 'start'|'ok'|'timeout'|'error', elapsedMs: number, error?: Error }) => void} [opts.onTransition]
   *   Optional observer fired before each console log line — lets tests assert
   *   on transitions without parsing console output.
   * @param {{ now(): number }} [opts.timerImpl]
   *   Injectable clock for timing — only `now()` is used by the ladder itself.
   *   Defaults to `Date.now`. (The action implementations may use the same
   *   clock for their setTimeout/clearTimeout, but that is the action's
   *   responsibility, not the ladder's.)
   */
  constructor({ rungs, onTransition, timerImpl } = {}) {
    if (!Array.isArray(rungs) || rungs.length === 0) {
      throw new TypeError('RecoveryLadder: rungs must be a non-empty array');
    }
    for (const r of rungs) {
      if (!r || typeof r.name !== 'string' || typeof r.action !== 'function') {
        throw new TypeError('RecoveryLadder: each rung needs { name, budgetMs, action }');
      }
    }

    this._rungs = rungs;
    this._onTransition = typeof onTransition === 'function' ? onTransition : null;
    this._now = timerImpl && typeof timerImpl.now === 'function'
      ? () => timerImpl.now()
      : () => Date.now();

    /** Single AbortController shared across all rungs in this ladder. */
    this._abort = new AbortController();
    /** Guards against double-run / post-cancel run. */
    this._started = false;
    /** Cached result so a second `await ladder.run()` returns the same outcome. */
    /** @type {Promise<LadderResult>|null} */
    this._runPromise = null;
  }

  /**
   * Walk every rung in order, returning early on the first success or on
   * cancellation. Idempotent: a second call returns the same promise.
   *
   * @returns {Promise<LadderResult>}
   */
  run() {
    if (this._runPromise) return this._runPromise;
    this._started = true;
    this._runPromise = this._runInternal();
    return this._runPromise;
  }

  /** @returns {Promise<LadderResult>} */
  async _runInternal() {
    /** @type {Error|null} */
    let lastError = null;

    for (const rung of this._rungs) {
      // If cancellation landed between rungs, surrender immediately.
      if (this._abort.signal.aborted) {
        this._log('cancelled');
        return { ok: false, cancelled: true };
      }

      const startedAt = this._now();
      this._emit({ rung: rung.name, phase: 'start', elapsedMs: 0 });
      this._log(`rung=${rung.name} start`);

      let ok = false;
      try {
        ok = await rung.action({ signal: this._abort.signal });
      } catch (err) {
        const elapsedMs = this._now() - startedAt;
        lastError = err instanceof Error ? err : new Error(String(err));
        this._emit({ rung: rung.name, phase: 'error', elapsedMs, error: lastError });
        this._log(`rung=${rung.name} error t=${fmtSeconds(elapsedMs)}s ${lastError.message}`);
        // Treat a thrown action the same as a "false" result: advance to the
        // next rung unless we were cancelled mid-action (signal already set).
        if (this._abort.signal.aborted) {
          this._log('cancelled');
          return { ok: false, cancelled: true, lastError };
        }
        continue; // eslint-disable-line no-continue
      }

      const elapsedMs = this._now() - startedAt;

      // If cancellation landed during a rung that still resolved truthy,
      // honour the cancel (caller asked us to stop) — we do NOT report a
      // success when the user explicitly cancelled. This is conservative but
      // correct: the spec's only cancellation path (`requestReconnect` mid-
      // ladder) wants Rung 3 to run regardless of any in-flight rung's
      // outcome (Task 11).
      if (this._abort.signal.aborted) {
        this._log('cancelled');
        return lastError
          ? { ok: false, cancelled: true, lastError }
          : { ok: false, cancelled: true };
      }

      if (ok) {
        this._emit({ rung: rung.name, phase: 'ok', elapsedMs });
        this._log(`rung=${rung.name} ok t=${fmtSeconds(elapsedMs)}s`);
        return { ok: true, rung: rung.name };
      }

      this._emit({ rung: rung.name, phase: 'timeout', elapsedMs });
      this._log(`rung=${rung.name} timeout t=${fmtSeconds(elapsedMs)}s`);
    }

    // Every rung returned false (or threw) without cancellation.
    return lastError
      ? { ok: false, exhausted: true, lastError }
      : { ok: false, exhausted: true };
  }

  /**
   * Signal every rung action to abort and stop the ladder. Safe to call
   * before, during, or after `run()`. The promise returned by `run()` will
   * resolve with `{ ok: false, cancelled: true }` once the in-flight action
   * has settled in response to the signal. Idempotent.
   */
  cancel() {
    if (this._abort.signal.aborted) return;
    this._abort.abort();
  }

  /** @returns {boolean} True after `cancel()` has been called. */
  get cancelled() { return this._abort.signal.aborted; }

  // ── internal helpers ──────────────────────────────────────────────────────

  _emit(transition) {
    if (!this._onTransition) return;
    try { this._onTransition(transition); } catch { /* observer must not break the ladder */ }
  }

  _log(tail) {
    // eslint-disable-next-line no-console
    console.log(`[CA] ladder: ${tail}`);
  }
}

/**
 * Wait until `predicate(observable.value)` is truthy OR `budgetMs` elapses
 * OR `signal` aborts. Resolves `true` on predicate match, `false` on budget
 * expiry / cancellation. Cleans up both the subscription and the timer on
 * every exit path.
 *
 * Used by the ConnectionAuthority's rung action closures to await peerHealth
 * / selfState transitions without each rung re-implementing the race.
 *
 * @template T
 * @param {{ subscribe(fn:(v:T)=>void): () => void, value: T }} observable
 *   An Observable-like object — see {@link ../connection/observable.js}. The
 *   subscriber is invoked synchronously with the current value once on
 *   subscribe (replay), then on every change. This helper relies on that
 *   replay semantics so a value that is already truthy at call-time wins
 *   without an additional event.
 * @param {(v: T) => boolean} predicate
 *   Called with each emitted value; returning truthy resolves the wait.
 * @param {number} budgetMs
 *   Maximum time to wait, in ms, measured by the injected timer.
 * @param {AbortSignal|undefined} signal
 *   If provided and `aborted`, the wait resolves `false` immediately. If
 *   abort fires later, the wait resolves `false` and cleans up.
 * @param {{ setTimeout(fn:()=>void, ms:number): any, clearTimeout(id:any): void }} timerImpl
 *   Injectable timer surface — typically the same FakeClock the authority is
 *   wired against in tests.
 * @returns {Promise<boolean>}
 */
export function awaitObservableOrTimeout(observable, predicate, budgetMs, signal, timerImpl) {
  return new Promise((resolve) => {
    // Short-circuit: caller already cancelled.
    if (signal && signal.aborted) {
      resolve(false);
      return;
    }

    let settled = false;
    let unsubscribe = null;
    let timerId = null;
    let onAbort = null;

    const cleanup = () => {
      if (unsubscribe) { try { unsubscribe(); } catch { /* swallow */ } unsubscribe = null; }
      if (timerId != null) { try { timerImpl.clearTimeout(timerId); } catch { /* swallow */ } timerId = null; }
      if (signal && onAbort) { try { signal.removeEventListener('abort', onAbort); } catch { /* swallow */ } onAbort = null; }
    };

    const settle = (value) => {
      if (settled) return;
      settled = true;
      cleanup();
      resolve(value);
    };

    // Arm the budget timer FIRST so the synchronous-replay subscribe below
    // can pre-empt it cleanly via cleanup() if the predicate is already true.
    timerId = timerImpl.setTimeout(() => settle(false), budgetMs);
    // In Node, setTimeout returns a Timeout object whose `unref()` removes
    // the timer from the event-loop "keep alive" set. We call it so a
    // ladder kicked from a test that doesn't await its resolution won't
    // hold the test process open for the full budget. FakeClock returns a
    // plain number — the typeof guard makes this a no-op there.
    if (timerId && typeof timerId.unref === 'function') timerId.unref();

    if (signal) {
      onAbort = () => settle(false);
      signal.addEventListener('abort', onAbort, { once: true });
    }

    // Subscribe LAST so the replay-on-subscribe semantics get a chance to
    // resolve us before the timer/abort listeners are running unnecessarily.
    // The Observable invokes `fn` synchronously with the current value during
    // subscribe(), so if the predicate is already true we settle here and
    // cleanup() tears the timer + listener back down before this function
    // returns.
    unsubscribe = observable.subscribe((v) => {
      if (settled) return;
      try {
        if (predicate(v)) settle(true);
      } catch {
        // Predicate threw — treat as no-match; budget timer remains armed.
      }
    });
  });
}

function fmtSeconds(ms) {
  return (ms / 1000).toFixed(1);
}

File: android/app/src/main/java/com/zaptransfer/android/connection/RecoveryLadder.kt
package com.zaptransfer.android.connection

import android.util.Log
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.supervisorScope
import java.util.concurrent.atomic.AtomicBoolean

private const val TAG = "RecoveryLadder"

/**
 * One step of a [RecoveryLadder].
 *
 * Mirror of the JS `{name, budgetMs, action}` object shape in
 * `extension/connection/recovery-ladder.js`. The action is invoked with a
 * [CoroutineScope] whose cancellation flips when [RecoveryLadder.cancel] is
 * called — actions MUST observe that cancellation (via a `withTimeout` /
 * `select` / cooperative `isActive` check inside their suspending helpers)
 * to tear down their subscriptions and timers cleanly. Otherwise
 * cancellation leaks.
 *
 * @property name Short identifier used in logs (`rung1`, `rung2`, ...).
 * @property budgetMs Informational — the action owns its own timeout via
 *           `kotlinx.coroutines.withTimeoutOrNull(budgetMs) { ... }`. The
 *           ladder logs per-rung elapsed time but does not enforce the
 *           budget itself.
 * @property action `suspend (CoroutineScope) -> Boolean` — returns `true`
 *           if the rung recovered, `false` if it timed out / surrendered.
 *           A thrown action advances to the next rung the same as `false`.
 *           The provided [CoroutineScope] is the rung's cancellation scope:
 *           cancelling it (via [RecoveryLadder.cancel]) propagates through
 *           any structured-concurrency helpers the action launches.
 */
data class RecoveryRung(
    val name: String,
    val budgetMs: Long,
    val action: suspend (CoroutineScope) -> Boolean,
)

/**
 * Result of a [RecoveryLadder.run] invocation.
 *
 * Mirror of the JS `LadderResult` shape `{ok, rung?, cancelled?, exhausted?,
 * lastError?}` from `extension/connection/recovery-ladder.js`. Modelled as a
 * sealed class so `when` branches are exhaustive at compile time.
 *
 *  - [Success]   The named rung recovered. `ok = true`.
 *  - [Cancelled] [RecoveryLadder.cancel] drove the exit. `ok = false`.
 *                `priorRung` is the name of the rung in flight when cancel
 *                landed (or `null` if cancel landed before any rung ran).
 *  - [Exhausted] Every rung returned false / threw. `ok = false`.
 *                `lastError` carries the most recent action throw (or null
 *                if every rung returned false rather than throwing).
 */
sealed class LadderResult {
    abstract val ok: Boolean

    /** A rung's action returned `true`. */
    data class Success(val rung: String) : LadderResult() {
        override val ok = true
    }

    /** [RecoveryLadder.cancel] was invoked during or before the ladder ran. */
    data class Cancelled(val priorRung: String?) : LadderResult() {
        override val ok = false
    }

    /** Every rung returned `false` or threw. */
    data class Exhausted(val lastError: Throwable? = null) : LadderResult() {
        override val ok = false
    }
}

/**
 * Sequential, cancelable runner for an ordered list of [RecoveryRung]s.
 *
 * Mirror of `extension/connection/recovery-ladder.js` (commit `dd87ae2`).
 * Used by [ConnectionAuthority] to recover from `peerHealth = FAILED` or
 * `selfState = RECONNECTING`. The ladder owns sequencing, logging, and
 * cancellation; each rung's [RecoveryRung.action] owns its own success-
 * signal subscription and budget-timer.
 *
 * ### Semantics
 *  - Rungs run strictly sequentially. The first action returning `true`
 *    ends the ladder with [LadderResult.Success].
 *  - An action returning `false` (or throwing) advances to the next rung.
 *  - If [cancel] is called, the in-flight rung's [CoroutineScope] is
 *    cancelled and the ladder resolves with [LadderResult.Cancelled] after
 *    the action settles. No further rungs are attempted.
 *  - If every rung fails without cancellation, the result is
 *    [LadderResult.Exhausted].
 *
 * ### Logging
 * Each phase emits a `[CA] ladder: ...` line so cross-platform diagnostic
 * dumps line up with Chrome's [extension/connection/recovery-ladder.js]:
 *  - `[CA] ladder: rung=N start`        on entry.
 *  - `[CA] ladder: rung=N ok t=X.Xs`    on success.
 *  - `[CA] ladder: rung=N timeout t=X.Xs` on action returning false.
 *  - `[CA] ladder: rung=N error t=X.Xs <message>` on throw.
 *  - `[CA] ladder: cancelled`           on cancel-driven exit.
 *
 * ### Threading
 * [run] is `suspend` and runs entirely on the caller's coroutine context.
 * The expected caller is [ConnectionAuthority]'s single-threaded worker
 * scope (`scope.launch { ladder.run() }`), which inherits the same
 * confinement guarantees as the rest of the authority. [cancel] is safe to
 * call from any thread.
 *
 * Spec: `docs/superpowers/specs/2026-05-02-connection-authority-design.md`
 *       §"Recovery ladder".
 */
class RecoveryLadder(private val rungs: List<RecoveryRung>) {

    init {
        require(rungs.isNotEmpty()) { "RecoveryLadder: rungs must be non-empty" }
    }

    /**
     * Cancellation latch. Flips on [cancel]; checked between rungs and read
     * after each rung settles to decide between [LadderResult.Success] /
     * [LadderResult.Cancelled] when the action and cancel race.
     */
    private val cancelled = AtomicBoolean(false)

    /**
     * Holds the in-flight rung's [Job] so [cancel] can propagate
     * cancellation into its [CoroutineScope]. Cleared between rungs.
     */
    @Volatile
    private var inFlightJob: Job? = null

    /**
     * Cached result for [run] idempotency. The [CompletableDeferred] is
     * completed once the first call to [run] settles, and any subsequent
     * call simply awaits it.
     */
    private val resultLatch = CompletableDeferred<LadderResult>()

    /** Guards against re-entering [runInternal] on a second [run] call. */
    private val started = AtomicBoolean(false)

    /** Wall-clock-equivalent — uses [System.nanoTime] to measure rung elapsed. */
    private fun now(): Long = System.nanoTime() / 1_000_000L

    /**
     * Walk every rung in order, returning early on the first success or on
     * cancellation. Idempotent — a second call awaits the same result.
     *
     * Returns one of [LadderResult.Success] / [LadderResult.Cancelled] /
     * [LadderResult.Exhausted]. Never throws.
     */
    suspend fun run(): LadderResult {
        if (started.compareAndSet(false, true)) {
            // First invocation — actually walk the rungs and settle the
            // shared latch with the outcome.
            val result = runInternal()
            resultLatch.complete(result)
            return result
        }
        // Second / Nth invocation — await the cached outcome.
        return resultLatch.await()
    }

    /**
     * Signal the in-flight rung to abort and stop the ladder.
     *
     * Safe to call before, during, or after [run]. The [run] suspension
     * resolves with [LadderResult.Cancelled] once the in-flight action has
     * settled in response to the cancellation. Idempotent — a second call
     * is a silent no-op.
     */
    fun cancel() {
        if (!cancelled.compareAndSet(false, true)) return
        // Cancel the in-flight rung's coroutine scope. Job.cancel() is a
        // no-op if the rung has already completed; the next-rung guard in
        // runInternal observes the cancelled flag and surrenders before
        // launching the next action.
        inFlightJob?.cancel()
    }

    /** True after [cancel] has been called. */
    val isCancelled: Boolean get() = cancelled.get()

    // ── internals ───────────────────────────────────────────────────────────

    private suspend fun runInternal(): LadderResult {
        var lastError: Throwable? = null
        var priorRung: String? = null

        for (rung in rungs) {
            // Cancellation between rungs: surrender immediately. priorRung
            // carries the most recently-attempted rung's name (or null if
            // cancel landed before any rung ran).
            if (cancelled.get()) {
                Log.i(TAG, "[CA] ladder: cancelled")
                return LadderResult.Cancelled(priorRung)
            }

            priorRung = rung.name
            val startedAt = now()
            Log.i(TAG, "[CA] ladder: rung=${rung.name} start")

            // Run the action inside a `supervisorScope` so we can cancel
            // its child Job via [cancel] without that cancellation
            // propagating up to the caller's scope. We hold a reference to
            // the child [Deferred] so [cancel] can target it cooperatively.
            // The supervisorScope block waits for the child to settle (or
            // for cancellation to drain) before returning.
            //
            // `supervisorScope` (vs `coroutineScope`) is the right choice
            // here because the rung's [Job] failure is a domain outcome,
            // not a parent-cancellation event — we want a thrown action to
            // bubble through `await()` as a normal exception we catch and
            // log, NOT to cancel the ladder's caller.
            val actionResult: ActionOutcome = supervisorScope {
                val deferred: Deferred<Boolean> = async { rung.action(this) }
                inFlightJob = deferred
                try {
                    ActionOutcome.Returned(deferred.await())
                } catch (ce: CancellationException) {
                    // The async call was cancelled — almost always via
                    // [cancel] flipping the latch and calling
                    // `inFlightJob?.cancel()`. The cancelled flag check
                    // below decides Cancelled vs. caller-cancellation
                    // rethrow.
                    ActionOutcome.Cancelled
                } catch (e: Throwable) {
                    ActionOutcome.Threw(e)
                } finally {
                    inFlightJob = null
                }
            }

            val elapsedMs = now() - startedAt

            when (actionResult) {
                is ActionOutcome.Cancelled -> {
                    // The rung observed cancellation. Per spec, this is a
                    // ladder-level cancel result regardless of any value
                    // the action might have computed.
                    Log.i(TAG, "[CA] ladder: cancelled")
                    return LadderResult.Cancelled(priorRung)
                }
                is ActionOutcome.Threw -> {
                    val err = (actionResult as ActionOutcome.Threw).error
                    lastError = err
                    Log.i(
                        TAG,
                        "[CA] ladder: rung=${rung.name} error t=${fmtSeconds(elapsedMs)}s ${err.message}",
                    )
                    // If cancel landed mid-throw, honour it.
                    if (cancelled.get()) {
                        Log.i(TAG, "[CA] ladder: cancelled")
                        return LadderResult.Cancelled(priorRung)
                    }
                    // Otherwise treat throw the same as `false`: advance.
                    continue
                }
                is ActionOutcome.Returned -> {
                    val ok = (actionResult as ActionOutcome.Returned).value
                    // Cancel-during-action ordering: if the user called
                    // cancel() mid-action and the action still resolved
                    // truthy, surrender the success and report cancelled.
                    // This is conservative but correct — the spec's
                    // requestReconnect path (Task 11, later) wants the new
                    // ladder to own the recovery state machine regardless.
                    if (cancelled.get()) {
                        Log.i(TAG, "[CA] ladder: cancelled")
                        return LadderResult.Cancelled(priorRung)
                    }
                    if (ok) {
                        Log.i(TAG, "[CA] ladder: rung=${rung.name} ok t=${fmtSeconds(elapsedMs)}s")
                        return LadderResult.Success(rung.name)
                    }
                    Log.i(TAG, "[CA] ladder: rung=${rung.name} timeout t=${fmtSeconds(elapsedMs)}s")
                    // Continue to the next rung.
                }
            }
        }

        // Every rung ran and surrendered (false or throw) without
        // cancellation — exhausted.
        return LadderResult.Exhausted(lastError)
    }

    /**
     * Tri-state outcome of a single rung's action invocation. We can't fold
     * this into [LadderResult] because the post-action loop branches on the
     * specific case (advance vs surrender vs return) before the ladder
     * settles its overall result.
     */
    private sealed class ActionOutcome {
        /** Action returned a boolean cleanly. */
        data class Returned(val value: Boolean) : ActionOutcome()

        /** Action threw a non-cancellation exception. */
        data class Threw(val error: Throwable) : ActionOutcome()

        /** Action's coroutine was cancelled (via [RecoveryLadder.cancel]). */
        object Cancelled : ActionOutcome()
    }

    private fun fmtSeconds(ms: Long): String {
        // One decimal place, matching the Chrome `(ms / 1000).toFixed(1)` output.
        val tenths = (ms + 50) / 100   // round to nearest tenth-of-a-second
        return "${tenths / 10}.${tenths % 10}"
    }
}
